### PR TITLE
package lists: Use nattka to parse bugs

### PR DESCRIPTION
**Pull request by thesamesam**
NATTkA is now used for most of the package lists on Bugzilla.
Use it to parse the lists because currently we cannot handle
package atoms with no version in keywording bugs.

Closes: https://github.com/gentoo/tatt/issues/66
Closes: https://github.com/gentoo/tatt/issues/65
Signed-off-by: Sam James (sam_c) <sam@cmpct.info>

### CODE REVIEW

**Comment by thesamesam** [comment]
This is still a work in progress.

**Comment by DerDakon** [comment]
Just to understand this correctly: your goal is to have the nattka subprocess parsing return packageList objects instead of just flat lists? That would be awesome.

**Comment by thesamesam** [comment]
Yep, that's the plan!

**Comment by thesamesam** [comment]
To test locally, I recommend adding the following to head of scripts/tatt.py:
```
import os.path
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
```

**Comment by thesamesam** [comment]
Please note that I need to fix manual package lists (-f).

**Comment by DerDakon** [comment]
It breaks when the bug fails sanity check. It also runs nattka even when the bug already has proper versions specified in it's package list.

**Comment by DerDakon** [comment]
Testcase: [bug 692324](https://bugs.gentoo.org/692324). Fails because of ppc (I don't understand it), but I want to build it for tatt.

**Comment by thesamesam** [comment]
My preference is to always run nattka rather than try to parse the list and figure out if we like it or not. Nattka is the de-facto format now and it's best if we support whatever it does. Does this sound okay to you? Or is there a simple way of checking if it's in an ok format for us to parse without it?

Regarding sanity, yeah, it'd be nice if we gave a warning that sanity check was failed but allow the user to proceed anyway. I'll look into this.

**Comment by DerDakon** [comment]
The easiest way is probably to try to run the current code, which should be rather fast. If that fails then run nattka.

**Comment by DerDakon** [comment]
The default for the repodir is nasty when nattka is used: the default is ~/gentoo-x86, which will not exist for most of us. Nattka will just fail then and does nothing. What does nattka do if we don't pass that dir if it does not exist? Best would be if it falls back to the normal portage directory, then we would be fine.

**Comment by DerDakon** [comment]
I now have added --ignore-sanity-check to the nattka commandline, which seems to make things much better.

**Comment by thesamesam** [comment]
Please let me know if you have any issues now.

File: scripts/tatt [hunk 1]
```diff
@@ -141,8 +142,7 @@ if options.infile:
     else:
         myJob.type="stable"
 
-    myJob.packageList = packageFinder.findPackages(packraw, targetarch)
-
```
Comment (DerDakon) [inline scripts/tatt:145]: This newline can be kept, it makes the code slightly clearer IMHO.

File: scripts/tatt [hunk 2]
```diff
@@ -207,9 +207,12 @@ if myJob.packageList is not None and len(myJob.packageList) > 0:
         if myJob.type == "keyword":
             targetarch = '~' + targetarch
 
+        output = myJob.packageList
```
Comment (DerDakon) [inline scripts/tatt:210]: Is this variable needed?

File: scripts/tatt [hunk 3]
```diff
@@ -207,9 +207,12 @@ if myJob.packageList is not None and len(myJob.packageList) > 0:
...
+
         for p in myJob.packageList:
             print("Found the following package atom : " + p.packageString())
             # check if the package already has the needed keywords
+            # TODO: Is this needed anymore? I imagine nattka won't emit empty ones.
```
Comment (DerDakon) [inline scripts/tatt:215]: Likely still needed for the local-file mode

File: tatt/packageFinder.py [hunk 4]
```diff
@@ -1,17 +1,28 @@
 """module for extracting packages from a package/architecture list """
-
+import subprocess
 from .gentooPackage import gentooPackage as gP
 
-def findPackages (s, arch):
+def findPackages (s, arch, repo, bugnum=False):
```
Comment (DerDakon) [inline tatt/packageFinder.py:5]: please also add the new params to the documentation

File: tatt/tool.py [hunk 5]
```diff
@@ -17,3 +19,15 @@ def idfun(x): return x
             seen[marker] = 1
             result.append(item)
     return result
+
+def get_repo_dir(repodir):
```
Comment (DerDakon) [inline tatt/tool.py:23]: Please factor this out into a first, independent commit. From our IRC discussion yesterday:

* remove the default ~/gentoo-x86 from the config
* if a repo dir is set in config this should be used, and if it does not exists that should be an error IMHO
* only go to the usual system defaults if the config hasn't set anything

File: scripts/tatt (hunk 2 above)
Comment (thesamesam) [reply scripts/tatt:210]: No!

File: scripts/tatt (hunk 3 above)
Comment (thesamesam) [reply scripts/tatt:215]: Agreed.

File: tatt/packageFinder.py (hunk 4 above)
Comment (thesamesam) [reply tatt/packageFinder.py:5]: Which documentation?

Comment (DerDakon) [reply tatt/packageFinder.py:5]: Lines 6-8.

Comment (thesamesam) [reply tatt/packageFinder.py:5]: I was a bit tired, I reckon. Done.

**Comment by mjeveritt** [comment]
Some good code updates here, fwiw, but I think this PR has grown a bit "out-of-scope" and needs splitting up @thesamesam . 

If 'upstream' doesn't want to adopt these changes, we need to pick a known-good fork for live AT :wink: :grin:

File: tatt/tool.py [hunk 6]
```diff
@@ -17,3 +19,21 @@ def idfun(x): return x
...
+            return repodir
+        else:
+            raise ValueError("Repo dir does not seem to be a directory")
+
+    # No path given in config
+    if os.path.isdir("/var/db/repos/gentoo/"):
+        print("Using /var/db/repos/gentoo/ as fallback")
+        return "/var/db/repos/gentoo"
+    elif os.path.isdir("/usr/portage/"):
+        print("Using /var/portage/ as fallback")
```
Comment (DerDakon) [inline tatt/tool.py:36]: no, "/usr/portage", not "/var/portage". Same in the commit message

**Review by DerDakon — CHANGES_REQUESTED** [review]
Besides the little /var/portage nitpick this seems all good now.

File: tatt/tool.py (hunk 6 above)
Comment (mjeveritt) [reply tatt/tool.py:36]: is he drunk-coding again?! :rofl: ..

Comment (thesamesam) [reply tatt/tool.py:36]: Just a typo..

**Comment by mjeveritt** [comment]
Ok now this needs a rebase too, with the relevant hunks merged into the correct commits/PRs .. ffs @thesamesam  :P lol

**Comment by thesamesam** [comment]
Not sure what you mean, I rebased it earlier based on master.

**Comment by mjeveritt** [comment]
> Not sure what you mean, I rebased it earlier based on master.

Yeah I'm not talking about replaying commits on different Heads, I'm talking Real rebasing, where you edit the commits, squash and such. I can see instances where you've made changes to previous commits in new commits .. that code can be 'fixed' in one change .. atomic commits and all that? I can probably give it an overhaul/further suggestion if you like .. (if that helps)

**Pull request closed without merging**